frame/pools/src/lib.rs

```diff
@@ -0,0 +1,1135 @@
+			// have exactly `MaxUnbonding` pools.
+			return self
+		}
+
+		//  I.E. if `MaxUnbonding` is 5 and current era is 10, we only want to retain pools 6..=10.
+		let newest_era_to_remove = current_era.saturating_sub(MaxUnbonding::<T>::get());
+
+		let eras_to_remove: Vec<_> = self
+			.with_era
+			.keys()
+			.cloned()
+			.filter(|era| *era <= newest_era_to_remove)
+			.collect();
+		for era in eras_to_remove {
+			if let Some(p) = self.with_era.remove(&era) {
+				self.no_era.points = self.no_era.points.saturating_add(p.points);
+				self.no_era.balance = self.no_era.balance.saturating_add(p.balance);
+			}
+		}
+
+		self
+	}
+}
+
+/// The maximum amount of eras an unbonding pool can exist prior to being merged with the
+/// `no_era	 pool. This is guaranteed to at least be equal to the staking `UnbondingDuration`. For
+/// improved UX [`Config::WithEraWithdrawWindow`] should be configured to a non-zero value.
+struct MaxUnbonding<T: Config>(PhantomData<T>);
+impl<T: Config> Get<u32> for MaxUnbonding<T> {
+	fn get() -> u32 {
+		// TODO: This may be too dangerous in the scenario bonding_duration gets decreased because
+		// we would no longer be able to decode `SubPoolsWithEra`, which uses `MaxUnbonding` as the
+		// bound
+		T::StakingInterface::bonding_duration() + T::WithEraWithdrawWindow::get()
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_system::{ensure_signed, pallet_prelude::*};
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(crate) trait Store)]
+	#[pallet::generate_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Weight information for extrinsics in this pallet.
+		// type WeightInfo: weights::WeightInfo;
+
+		/// The nominating balance.
+		type Currency: Currency<Self::AccountId>;
+
+		// Infallible method for converting `Currency::Balance` to `U256`.
+		type BalanceToU256: Convert<BalanceOf<Self>, U256>;
+
+		// Infallible method for converting `U256` to `Currency::Balance`.
+		type U256ToBalance: Convert<U256, BalanceOf<Self>>;
+
+		/// The interface for nominating.
+		type StakingInterface: StakingInterface<
+			Balance = BalanceOf<Self>,
+			AccountId = Self::AccountId,
+			LookupSource = <Self::Lookup as StaticLookup>::Source,
+		>;
+
+		/// The amount of eras a `SubPools::with_era` pool can exist before it gets merged into the
+		/// `SubPools::no_era` pool. In other words, this is the amount of eras a delegator will be
+		/// able to withdraw from an unbonding pool which is guaranteed to have the correct ratio of
+		/// points to balance; once the `with_era` pool is merged into the `no_era` pool, the ratio
+		/// can become skewed due to some slashed ratio getting merged in at some point.
+		type WithEraWithdrawWindow: Get<u32>;
```

Comment:
```suggestion
		type PostUnbondingPoolsWindow: Get<u32>;
```

The term `with_era` is an internal detail and not really expressive of what it is supposed to meant.

Reply:
f8430c299c310bf71b048c5fdad55e6301feb335